helper/schema/schema.go

```diff
@@ -1731,12 +1731,25 @@ func (m schemaMap) validatePrimitive(
 		}
 		decoded = n
 	case TypeInt:
-		// Verify that we can parse this as an int
-		var n int
-		if err := mapstructure.WeakDecode(raw, &n); err != nil {
-			return nil, []error{fmt.Errorf("%s: %s", k, err)}
+		switch {
+		case isProto5():
+			// We need to verify the type precisely, because WeakDecode will
+			// decode a float as an integer.
+
+			// the config shims only use int for integral number values
+			if v, ok := raw.(int); ok {
+				decoded = v
+			} else {
+				return nil, []error{fmt.Errorf("%s: expected int, got %[2]T(%[2]v)", k, raw)}
```

Comment:
It's a shame that we have to handle this so late here, when we've already shimmed everything, because `cty` has a nicer error message for this situation, but I guess we can live for now and improve it in the future SDK.

With that said, we _could_ try to mimic `cty`'s error message style here right now, which is to say "must be a whole number" rather than "expected int" if the given value is a float. "int" and "float64" are Go types, not Terraform language types, so to include them in a user-facing error message is likely to cause confusion.

I also wondered here if we can be sure that a whole number from the configuration will always appear as an `int` here. Is there ever a case where a float with no fractional part would be used instead? It looks like the shim prefers to use int where possible:

https://github.com/hashicorp/terraform/blob/28b2383eacaefaa88dd67afdf1d3af28e21900c0/config/hcl2shim/values.go#L151-L162

...so seems fine, I guess.

Reply:
Yes, I'd rather get this earlier, but I don't think we have any choice since the int/float differentiation is only in `helper/schema`.  I can see this being a common request for property validation in a new SDK too, so I don't know if we want to push this down into configschema somehow later on. 

Good idea on the error message, I'll change that before merging.